File: DRISHTI_portable/matlab/module1_quality_gate.m
function [decision, finalImg, report] = module1_quality_gate(img)
% =========================================================================
% DRISHTI - MODULE 1: THE TRUST GATE (MATLAB version)
% =========================================================================
% Quality assessment (focus / illumination / field of view) + the
% ACCEPT / ENHANCE / REJECT decision, identical logic to the Python
% prototype so results can be cross-checked.
%
% Inputs : img   - RGB fundus image (uint8, any size)
% Outputs: decision - 'ACCEPT' | 'ACCEPT_AFTER_ENHANCE' | 'REJECT'
%          finalImg - image to send to Module 2 (original or enhanced)
%          report   - struct with all quality metrics
% Requires: Image Processing Toolbox
% =========================================================================

report.mask = getRetinaMask(img);

[report.focus, report.laplacianVar] = focusScore(img, report.mask);
[report.illumination, report.meanBrightness] = illuminationScore(img, report.mask);
[report.fieldOfView, report.retinaCoverage] = fovScore(report.mask);

report.qualityScore = 0.40*report.focus + 0.30*report.illumination + 0.30*report.fieldOfView;

if report.qualityScore >= 0.80
    decision = 'ACCEPT';
    finalImg = img;
    report.reason = 'Image quality is good. Proceed to analysis.';
elseif report.qualityScore >= 0.50
    % --- ENHANCE and re-check ---
    enhanced = enhanceImage(img);
    rep2 = struct();
    rep2.mask = report.mask;
    [rep2.focus, rep2.laplacianVar] = focusScore(enhanced, rep2.mask);
    [rep2.illumination, rep2.meanBrightness] = illuminationScore(enhanced, rep2.mask);
    [rep2.fieldOfView, rep2.retinaCoverage] = fovScore(rep2.mask);
    rep2.qualityScore = 0.40*rep2.focus + 0.30*rep2.illumination + 0.30*rep2.fieldOfView;

    % Enhancement can fix brightness but NEVER blur (information is lost)
    if rep2.qualityScore >= 0.62 && rep2.focus >= 0.30
        decision = 'ACCEPT_AFTER_ENHANCE';
        finalImg = enhanced;
        report = rep2;
        report.enhanced = true;
        report.reason = 'Borderline image rescued by CLAHE enhancement.';
    else
        decision = 'REJECT';
        finalImg = img;
        report.reason = ['RECAPTURE NEEDED: enhancement could not rescue ' ...
                         'this image. Ask the health worker to retake the photo.'];
    end
else
    decision = 'REJECT';
    finalImg = img;
    report.reason = buildRejectReason(report);
end
end

% -------------------------------------------------------------------------
function mask = getRetinaMask(img)
% Binary mask of the retina (bright circle) inside the camera frame
gray = rgb2gray(img);
mask = gray > 15;                                   % retina brighter than bg
mask = imclose(mask, strel('disk', 8));            % fill small holes
mask = imopen(mask, strel('disk', 8));             % remove specks
mask = bwareaopen(mask, numel(mask)*0.05);         % keep main blob only
mask = uint8(mask)*255;
end

% -------------------------------------------------------------------------
function [score, lapVar] = focusScore(img, mask)
% Sharpness via Laplacian variance + Tenengrad, measured well INSIDE the
% retina (eroded mask) so borders/eyelids cannot inflate the score.
gray = im2single(rgb2gray(img));
inner = imerode(logical(mask), strel('disk', 12));
if ~any(inner(:))
    inner = logical(mask);
end

lap = imfilter(gray, fspecial('laplacian', 0.2));
lapVar = var(lap(inner));

% NOTE: MATLAB cannot index an expression directly, so we need a temporary:
[gx, gy] = imgradientxy(gray, 'sobel');
edgeEnergy = gx.^2 + gy.^2;
tenengrad = mean(edgeEnergy(inner));

s1 = 1 - exp(-lapVar/100);
s2 = 1 - exp(-tenengrad/2200);
score = 0.6*s1 + 0.4*s2;
if lapVar < 12
    score = min(score, 0.30);   % essentially featureless -> cannot rescue
end
end

% -------------------------------------------------------------------------
function [score, meanBright] = illuminationScore(img, mask)
% Average brightness + 8x8-grid uniformity (cheap cameras light unevenly)
gray = rgb2gray(img);
meanBright = mean(gray(logical(mask)));

if meanBright >= 60 && meanBright <= 200
    brightS = 1.0;
elseif meanBright < 60
    brightS = max(0, meanBright/60);
else
    brightS = max(0, (255-meanBright)/55);
end

% 8x8 grid uniformity
[H, W] = size(gray);
ch = floor(H/8); cw = floor(W/8);
means = zeros(1, 64); k = 0;
for i = 1:8
    for j = 1:8
        cellImg = gray((i-1)*ch+1:i*ch, (j-1)*cw+1:j*cw);
        cellMsk = mask((i-1)*ch+1:i*ch, (j-1)*cw+1:j*cw);
        if mean(cellMsk(:) > 0) > 0.4
            k = k + 1;
            means(k) = mean(cellImg(cellMsk > 0));
        end
    end
end
means = means(1:k);
if numel(means) < 4
    uniformS = 0;
else
    uniformS = exp(-std(means)/45);
end
score = 0.55*brightS + 0.45*uniformS;
end

% -------------------------------------------------------------------------
function [score, coverage] = fovScore(mask)
% Coverage of frame + fill of bounding box + circularity of the retina blob.
% (The Python prototype uses the same three sub-scores.)
m = logical(mask);
coverage = mean(m(:));
covS = min(max((coverage - 0.40)/(0.75 - 0.40), 0), 1);

stats = regionprops(m, 'Area', 'BoundingBox', 'Perimeter');
if ~isempty(stats) && stats(1).Area > 0
    area = stats(1).Area;
    bb = stats(1).BoundingBox;
    per = stats(1).Perimeter;
    fillS = min(area/(bb(3)*bb(4))/0.75, 1);        % healthy retina fills its bbox
    circ = 4*pi*area/max(per^2, eps);               % circularity: 1 = perfect circle
    shapeS = min(circ/0.85, 1);
else
    fillS = 0.5;
    shapeS = 0.5;
end
score = 0.45*covS + 0.30*fillS + 0.25*shapeS;
end

% -------------------------------------------------------------------------
function out = enhanceImage(img)
% CLAHE on the L channel + non-local-means denoising + gamma correction.
% (Same enhancement stack as the Python prototype and our PPT slide 6.)
lab = rgb2lab(im2single(img));

% --- CLAHE on lightness (uint8 path = safe across MATLAB versions) ---
L8 = im2uint8(lab(:,:,1)/100);                     % L in [0,1] -> uint8 [0,255]
L8 = adapthisteq(L8, 'ClipLimit', 0.02, 'NumTiles', [8 8]);
lab(:,:,1) = single(L8)/255*100;                   % back to Lab L scale
out = lab2rgb(lab);

% --- edge-preserving denoising (non-local means, as named in our PPT) ---
try
    out = imnlmfilt(out);
catch
    out = imgaussfilt(out, 0.8);                   % fallback: gentle blur
end

% --- gamma correction for very dark / over-exposed images ---
g = rgb2gray(out);
vals = g(g > 0.06);
if isempty(vals)
    mb = 0.5;
else
    mb = mean(vals);
end
if mb < 0.28
    out = out.^0.6;                                % brighten dark images
elseif mb > 0.75
    out = out.^1.4;                                % darken over-exposed images
end
out = im2uint8(min(max(out, 0), 1));
end

% -------------------------------------------------------------------------
function reason = buildRejectReason(report)
% A SPECIFIC recapture message for the health worker (field usability!)
reason = 'RECAPTURE NEEDED: ';
if report.focus < 0.35
    reason = [reason 'image too blurry - ask patient to hold still and refocus. '];
end
if report.illumination < 0.35
    reason = [reason 'bad lighting - check camera flash / room lighting. '];
end
if report.fieldOfView < 0.35
    reason = [reason 'retina not fully captured - realign camera closer to pupil.'];
end
end

File: DRISHTI_portable/src/module1_quality_gate.py
"""
============================================================================
DRISHTI - MODULE 1: THE TRUST GATE (Image Quality Assessment + Enhancement)
============================================================================
WHAT THIS MODULE DOES (in plain words):
    A retina photo taken in a village health centre is often blurry, too
    dark, or badly framed. A doctor would refuse to read such a photo.
    Our AI must do the same -> otherwise it will make WRONG predictions
    on bad images and people will lose trust.

    So before ANY analysis, we give every image a "quality report card"
    with 3 marks (like a school report):

        1. FOCUS        - is the image sharp? (can we see tiny blood vessels?)
        2. ILLUMINATION - is the brightness good and even?
        3. FIELD OF VIEW- did the camera capture enough of the retina?

    Then we take ONE of three decisions (this is the "TRUST GATE"):
        ACCEPT  (score > 0.80)  -> send image to Module 2 for analysis
        ENHANCE (score 0.5-0.80)-> fix the image first, then re-check
        REJECT  (score < 0.50)  -> tell the health worker to take a NEW photo,
                                   WITH the specific reason why it failed

MEDICAL BACKGROUND:
    * Fundus photo = photo of the back of the eye (the "retina")
    * CLAHE = Contrast Limited Adaptive Histogram Equalization.
      A famous technique that brightens dark areas of an image locally
      without over-brightening the already-bright areas.
============================================================================
"""

import cv2
import numpy as np


# --------------------------------------------------------------------------
# STEP 0: Find the retina (the useful circular area) inside the photo
# --------------------------------------------------------------------------
def get_retina_mask(img):
    """
    The camera photo is a black square with a bright circle (the retina)
    inside it. We create a binary mask (white = retina, black = background)
    so all our measurements are done ONLY on the retina, not the black area.

    Returns: mask (uint8, 255 = retina, 0 = background)
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

    # Retina pixels are brighter than the black background (threshold = 15)
    _, mask = cv2.threshold(gray, 15, 255, cv2.THRESH_BINARY)

    # Morphological operations = cleaning up the mask:
    #   CLOSE = fill small holes,  OPEN = remove small noise dots
    kernel = np.ones((15, 15), np.uint8)
    mask = cv2.morphologyEx(mask, cv2.MORPH_CLOSE, kernel)
    mask = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)

    # Keep only the BIGGEST bright region (= the retina itself)
    num_labels, labels, stats, _ = cv2.connectedComponentsWithStats(mask, connectivity=8)
    if num_labels <= 1:
        return mask  # nothing found, return as-is
    largest = 1 + np.argmax(stats[1:, cv2.CC_STAT_AREA])  # skip label 0 (bg)
    mask = np.where(labels == largest, 255, 0).astype(np.uint8)
    return mask


# --------------------------------------------------------------------------
# QUALITY METRIC 1: FOCUS (sharpness)
# --------------------------------------------------------------------------
def focus_score(img, mask):
    """
    BLURRY image = we cannot see microaneurysms (tiny red dots that are the
    EARLIEST sign of diabetes). A blurry image is medically USELESS.

    How we measure sharpness:
      * Laplacian filter detects edges (vessel walls). In a sharp image
        there are many strong edges -> high variance of the Laplacian.
      * Tenengrad (Sobel gradients) measures edge energy similarly.
    We combine both and squash the result into 0..1.

    IMPORTANT: we measure well INSIDE the retina (eroded mask). Otherwise a
    sharp eyelid edge or the black photo border would falsely raise the
    score of an otherwise blurry image.
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY).astype(np.float32)

    inner = cv2.erode(mask, np.ones((25, 25), np.uint8))   # away from borders
    if inner.sum() == 0:
        inner = mask

    # --- Laplacian variance (classic blur detector) ---
    lap = cv2.Laplacian(gray, cv2.CV_32F, ksize=3)
    lap_var = lap[inner > 0].var()

    # --- Tenengrad edge energy ---
    gx = cv2.Sobel(gray, cv2.CV_32F, 1, 0, ksize=3)
    gy = cv2.Sobel(gray, cv2.CV_32F, 0, 1, ksize=3)
    tenengrad = np.mean((gx ** 2 + gy ** 2)[inner > 0])

    # Squash to 0..1 (higher = sharper). Constants tuned for fundus photos.
    s1 = 1.0 - np.exp(-lap_var / 100.0)
    s2 = 1.0 - np.exp(-tenengrad / 2200.0)
    score = float(0.6 * s1 + 0.4 * s2)
    # Hard cap: if there is essentially NO edge energy, the image is blurred
    # (or completely featureless) - no enhancement can ever rescue it.
    if lap_var < 12:
        score = min(score, 0.30)
    return score, float(lap_var)


# --------------------------------------------------------------------------
# QUALITY METRIC 2: ILLUMINATION (brightness + evenness)
# --------------------------------------------------------------------------
def illumination_score(img, mask):
    """
    Too dark -> we miss lesions.  Too bright -> everything is washed out.
    Uneven lighting (common with cheap handheld cameras) -> one side of the
    photo looks different from the other side.

    We split the retina into an 8x8 grid (like a chess board) and check:
      * the average brightness of the whole retina (must be moderate)
      * how different the 64 grid cells are (must be similar = "uniform")
    """
    gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
    h, w = gray.shape

    # --- Average brightness inside the retina only ---
    mean_bright = float(gray[mask > 0].mean())

    # Ideal brightness ~ 60..200. Outside this range, penalise gradually.
    if 60 <= mean_bright <= 200:
        bright_s = 1.0
    elif mean_bright < 60:
        bright_s = max(0.0, mean_bright / 60.0)      # too dark
    else:
        bright_s = max(0.0, (255.0 - mean_bright) / 55.0)  # too bright

    # --- Uniformity: 8x8 grid brightness comparison ---
    cell_h, cell_w = h // 8, w // 8
    cell_means = []
    for i in range(8):
        for j in range(8):
            cell = gray[i * cell_h:(i + 1) * cell_h, j * cell_w:(j + 1) * cell_w]
            cell_mask = mask[i * cell_h:(i + 1) * cell_h, j * cell_w:(j + 1) * cell_w]
            if cell_mask.mean() > 0.4:  # only cells that are mostly retina
                cell_means.append(cell[cell_mask > 0].mean())
    if len(cell_means) < 4:
        uniform_s = 0.0   # we could not even measure -> suspicious image
    else:
        spread = float(np.std(cell_means))
        uniform_s = float(np.exp(-spread / 45.0))  # low spread -> high score

    return float(0.55 * bright_s + 0.45 * uniform_s), mean_bright


# --------------------------------------------------------------------------
# QUALITY METRIC 3: FIELD OF VIEW (did we capture enough retina?)
# --------------------------------------------------------------------------
def fov_score(img, mask):
    """
    If only a tiny sliver of retina was captured, we cannot grade DR.
    We check:
      * coverage: how much of the photo frame is retina (healthy fundus
        photos cover ~75-90% of the frame)
      * fill ratio: retina area vs its own bounding box. A healthy retina
        is one big round/oval blob (fill ~0.75). An eyelid or eyelash
        cutting into the photo leaves the bounding box mostly empty.
      * shape: how round the blob is
    """
    coverage = float((mask > 0).mean())

    # Coverage score: full marks at 0.75+, zero below 0.40
    cov_s = float(np.clip((coverage - 0.40) / (0.75 - 0.40), 0.0, 1.0))

    cnts, _ = cv2.findContours(mask, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)
    shape_s, fill_s = 0.5, 0.5
    if cnts:
        c = max(cnts, key=cv2.contourArea)
        area = cv2.contourArea(c)
        (cx, cy), radius = cv2.minEnclosingCircle(c)
        circle_area = np.pi * radius * radius
        shape_s = float(area / circle_area) if circle_area > 0 else 0.5
        x, y, bw, bh = cv2.boundingRect(c)
        fill_s = float(np.clip(area / (bw * bh + 1e-6) / 0.75, 0.0, 1.0))

    score = float(0.45 * cov_s + 0.30 * fill_s + 0.25 * min(1.0, shape_s / 0.85))
    return score, coverage


# --------------------------------------------------------------------------
# THE COMPLETE QUALITY REPORT (all 3 metrics -> one decision)
# --------------------------------------------------------------------------
def assess_quality(img):
    """
    Runs all 3 checks and returns a full quality report as a dictionary.
    Final score is a weighted average:  focus 40%, illumination 30%, FOV 30%.
    """
    mask = get_retina_mask(img)

    focus, lap_var = focus_score(img, mask)
    illum, mean_bright = illumination_score(img, mask)
    fov, coverage = fov_score(img, mask)

    score = float(0.40 * focus + 0.30 * illum + 0.30 * fov)

    # ---- TRUST GATE DECISION ----
    if score >= 0.80:
        decision, reason = "ACCEPT", "Image quality is good. Proceed to analysis."
    elif score >= 0.50:
        decision, reason = "ENHANCE", "Borderline quality. Applying enhancement and re-checking."
    else:
        # Build a SPECIFIC recapture message for the health worker:
        reasons = []
        if focus < 0.35:
            reasons.append("image too blurry - ask patient to hold still and refocus")
        if illum < 0.35:
            reasons.append("bad lighting - check camera flash / room lighting")
        if fov < 0.35:
            reasons.append("retina not fully captured - realign camera closer to pupil")
        if not reasons:
            reasons.append("overall quality too low - retake the photo")
        decision = "REJECT"
        reason = "RECAPTURE NEEDED: " + "; ".join(reasons)

    return {
        "quality_score": round(score, 3),
        "focus": round(focus, 3),
        "illumination": round(illum, 3),
        "field_of_view": round(fov, 3),
        "laplacian_variance": round(lap_var, 1),
        "mean_brightness": round(mean_bright, 1),
        "retina_coverage": round(coverage, 3),
        "decision": decision,
        "reason": reason,
        "mask": mask,
    }


# --------------------------------------------------------------------------
# ENHANCEMENT: rescue borderline images (CLAHE + denoising)
# --------------------------------------------------------------------------
def enhance_image(img):
    """
    For "borderline" images we apply 3 rescue steps (all from our PPT):
      1. CLAHE on the L (lightness) channel -> local contrast boost
      2. Non-local means denoising -> removes camera sensor noise
      3. Gentle gamma correction if image was too dark/bright
    Returns the enhanced image.
    """
    # --- Step 1: CLAHE on lightness channel of LAB colour space ---
    lab = cv2.cvtColor(img, cv2.COLOR_BGR2LAB)
    l_channel, a_channel, b_channel = cv2.split(lab)
    clahe = cv2.createCLAHE(clipLimit=2.5, tileGridSize=(8, 8))  # 8x8 tiles
    l_channel = clahe.apply(l_channel)
    enhanced = cv2.cvtColor(cv2.merge([l_channel, a_channel, b_channel]), cv2.COLOR_LAB2BGR)

    # --- Step 2: Non-local means denoising (keeps edges, removes noise) ---
    enhanced = cv2.fastNlMeansDenoisingColored(enhanced, None, 5, 5, 7, 21)

    # --- Step 3: gamma correction for very dark images ---
    gray = cv2.cvtColor(enhanced, cv2.COLOR_BGR2GRAY)
    mean_bright = float(gray[gray > 15].mean()) if (gray > 15).any() else 0.0
    if mean_bright < 70:
        gamma = 0.6   # brighten dark images (gamma < 1 brightens)
        table = np.array([((i / 255.0) ** gamma) * 255 for i in range(256)]).astype("uint8")
        enhanced = cv2.LUT(enhanced, table)
    elif mean_bright > 190:
        gamma = 1.4   # darken over-exposed images
        table = np.array([((i / 255.0) ** gamma) * 255 for i in range(256)]).astype("uint8")
        enhanced = cv2.LUT(enhanced, table)

    return enhanced


def quality_gate(img, verbose=True):
    """
    THE MAIN FUNCTION of Module 1. Give it any fundus image.
    Returns: (decision, final_image, quality_report)

    Logic:
        1. Measure quality.
        2. If ENHANCE -> enhance, then measure AGAIN (did we rescue it?)
        3. If still < 0.5 after enhancement -> REJECT (honest system!)
    """
    report = assess_quality(img)
    decision = report["decision"]
    final_img = img

    if decision == "ENHANCE":
        enhanced = enhance_image(img)
        report2 = assess_quality(enhanced)
        if verbose:
            print(f"   [TRUST GATE] borderline ({report['quality_score']:.2f}) "
                  f"-> enhanced -> new score {report2['quality_score']:.2f}")
        # Enhancement can fix brightness, but NEVER blur (information lost).
        # So we require BOTH: decent overall score AND recoverable focus.
        if report2["quality_score"] >= 0.62 and report2["focus"] >= 0.30:
            decision, final_img, report = "ACCEPT_AFTER_ENHANCE", enhanced, report2
            report["decision"] = decision
            report["enhanced"] = True
        else:
            why_blur = "" if report2["focus"] >= 0.30 else " image is too blurry to enhance -"
            decision, report = "REJECT", report
            report["decision"] = "REJECT"
            report["reason"] = (f"RECAPTURE NEEDED:{why_blur} enhancement could not "
                                "rescue this image (" + report["reason"].lower() + ")")

    if verbose:
        print(f"   [TRUST GATE] decision = {decision}  |  score = {report['quality_score']:.2f}")
    return decision, final_img, report
